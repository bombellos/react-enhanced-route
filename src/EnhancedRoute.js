import React from 'react';
import PropTypes from 'prop-types';
import { Route, Redirect } from 'react-router-dom';

export const EnhancedRoute = (props) => {

  const {component: Component, layout: Layout, redirect, auth: isAuthorized, ...rest} = props;

  /*
   If `auth <Boolean>` and `layout <ReactElement>` have not been provided
   as EnhancedRoute attribute then we return default
   react-router-dom Route.
   */
  if (!props.hasOwnProperty('auth') && !props.hasOwnProperty('layout'))
    return <Route {...props} />;

  /*
   Depending on whether we passed `layout <ReactElement>`
   attribute or not, we define template.
   */
  const template = Layout ? <Layout><Component /></Layout> : <Component/>;

  /*
   If `layout <ReactElement>` has been provided and
   `auth <Boolean>` has not been provided we return
   Route wrapped in specified layout.
   */
  if (props.hasOwnProperty('layout') && !props.hasOwnProperty('auth'))
    return <Route {...rest} component={() => template}/>;

  /*
   If we are not authorized we redirect to specified path
   or render default Unauthorized page.
   */
  if (!isAuthorized) {
    return redirect ? <Redirect to={{pathname: redirect}}/> :
        <Route render={() => <div>Unauthorized</div>}/>;
  }

  /*
   If we reached this far it means we are authorized
   and we render protected route.
   */
  return <Route {...rest} render={() => template}/>;
};

EnhancedRoute.propTypes = {
  component: PropTypes.func.isRequired,
  redirect: PropTypes.string,
  auth: PropTypes.bool,
  layout: PropTypes.func,
};