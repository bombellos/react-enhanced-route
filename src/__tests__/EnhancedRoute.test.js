import React from 'react';
import {BrowserRouter as Router} from 'react-router-dom';
import { EnhancedRoute as Route } from '../EnhancedRoute';

describe('EnhancedRoute test', () => {

  const SideBar = () => <div>Sidebar</div>;
  const Header = () => <div>Header</div>;
  const MainLayout = (props) => <div><Header/><SideBar/>{props.children}</div>;

  const AdminPage = () => <div>Admin Page</div>;
  const PublicPage = () => <div>Public Page</div>;

  let isAuthorized;

  beforeEach(() => {
    isAuthorized = true;
  });

  it('should render default route with provided component', () => {

    const element = render(<Router><Route component={PublicPage}/></Router>);

    expect(element).toMatchSnapshot();
  });

  it('should render route with provided layout and child component', () => {

    const element = render(<Router><Route layout={MainLayout} component={PublicPage}/></Router>);

    expect(element).toMatchSnapshot();
  });

  it('should render and allow access to protected route with provided component', () => {

    const element = render(<Router><Route auth={isAuthorized} component={AdminPage}/></Router>);

    expect(element).toMatchSnapshot();
  });

  it('should render and allow access to protected route with provided component and layout', () => {

    const element = render(<Router><Route auth={isAuthorized} layout={MainLayout} component={AdminPage}/></Router>);

    expect(element).toMatchSnapshot();
  });

  it('should redirect to provided page when accessing protected route', () => {

    isAuthorized = false;

    const element = render(<Router><Route auth={isAuthorized} redirect={'/login'} component={AdminPage}/></Router>);

    expect(element).toMatchSnapshot();
  });

  it('should render default unauthorized page when accessing protected route', () => {

    isAuthorized = false;

    const element = render(<Router><Route auth={isAuthorized} component={AdminPage}/></Router>);

    expect(element).toMatchSnapshot();
  });

});