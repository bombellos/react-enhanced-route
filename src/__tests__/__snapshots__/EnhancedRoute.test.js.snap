// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirect to provided page when accessing protected route 1`] = `null`;

exports[`render and allow access to protected route with provided component 1`] = `
<div>
  Admin Page
</div>
`;

exports[`render and allow access to protected route with provided component and layout 1`] = `
<div>
  <div>
    Header
  </div>
  <div>
    Sidebar
  </div>
  <div>
    Admin Page
  </div>
</div>
`;

exports[`render default route with provided component 1`] = `
<div>
  Public Page
</div>
`;

exports[`render default unauthorized page when accessing protected route 1`] = `
<div>
  Unauthorized
</div>
`;

exports[`render route with provided layout and child component 1`] = `
<div>
  <div>
    Header
  </div>
  <div>
    Sidebar
  </div>
  <div>
    Public Page
  </div>
</div>
`;
